File: 300/solution.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> dp = {nums[0]};

        for(int i=0; i<nums.size(); ++i){
            if(nums[i] > dp.back())
                dp.push_back(nums[i]);
            else
                dp[binarysearch(dp, nums[i])] = nums[i];
        }
        return dp.size();
    }
    int binarysearch(vector<int>& dp, int& target){
        int start = 0, end = dp.size()-1;
        
        while(start < end){
            int mid = (start+end)/2;
            if(dp[mid] < target)
                start = mid+1;
            else
                end = mid;
        }
        return end;
    }
};

File: 300/solution.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        res = []

        def binary_search(res, n):
            left = 0
            right = len(res)-1
            while left <= right:
                mid = (left+right)//2
                if res[mid] == n:
                    return mid
                elif res[mid] > n:
                    right = mid-1
                else:
                    left = mid+1

            return left

        for i in range(len(nums)):
            if not res or nums[i]>res[-1]:
                res.append(nums[i])
            else:
                temp = binary_search(res, nums[i])
                res[temp] = nums[i]

        return len(res)
